db/commitlog/commitlog.cc

```diff
@@ -1428,11 +1428,19 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
 
     ++_new_counter;
 
-    // don't increase reserve count if we are at max, or we would go over disk limit. 
-    if (_reserve_segments.empty() && (_reserve_segments.max_size() < cfg.max_reserve_segments) && (totals.total_size_on_disk + max_size) <= max_disk_size) {
-        _reserve_segments.set_max_size(_reserve_segments.max_size() + 1);
-        clogger.debug("Increased segment reserve count to {}", _reserve_segments.max_size());
+    if (_reserve_segments.empty()) {        
+        // don't increase reserve count if we are at max, or we would go over disk limit. 
+        if (_reserve_segments.max_size() < cfg.max_reserve_segments && (totals.total_size_on_disk + max_size) <= max_disk_size) {
+            _reserve_segments.set_max_size(_reserve_segments.max_size() + 1);
+            clogger.debug("Increased segment reserve count to {}", _reserve_segments.max_size());
+        }
+        // if we have no reserve and we're above/at limits, make background task a little more eager.
+        if (!_shutdown && totals.total_size_on_disk >= disk_usage_threshold) {
+            _timer.cancel();
+            _timer.arm(std::chrono::milliseconds(0));
+        }
```

Comment:
I don't know if this is needed, especially if we move the logic to flush from a timer to segment close.

Are we preallocating far enough ahead? Perhaps we should make preallocation completely independent of the rest of commitlog; launch a fiber to preallocate until we reach our limit, then stop.

Reply:
Preallocation is done "on-demand" as soon as reserve is empty. This is to try to enforce flushing of memtables to allow this to finish.

Reply:
I think we should launch preallocation during startup (unless, perhaps, --developer-mode). This is to avoid stalls when inserting the first terabyte of data.

What about dropping that timer? It might simplify things.

Reply:
We do. It is part of startup. 
Timer just says we want to encourage memtable flushes now please.